[cmd][api] Exporter/Importer command for CodeChecker cmd

jay24rajput opened the pull request:
## Exporter
An exclusive command for getting the comments and review statuses for a particular run, in JSON format from CodeChecker Server.  
Multiple run names can be provided for exporting the data simultaneously.
If no run name is provided, by default, data is exported for all the runs

### Usage
```sh
CodeChecker cmd export -n <run_name-1> <run_name-2>...  2>/dev/null | python -m json.tool > <run_name>.json
```

## Importer
A command for importing comments and review statuses from a json file into codechecker server

### Usage
```sh
CodeChecker cmd import -i <path_to_json_file>.json
```

Fixes #3076

csordasmarton commented on web/api/report_server.thrift:
```diff
@@ -724,4 +729,14 @@ service codeCheckerDBAccess {
+  
+  // Export data from the server
+  // PERMISSION: PRODUCT_ACCESS
+  ExportData exportData(1: i64 runId)
```
Unfortunately this API function does not allow to filter by multiple run names which I think should be supported. For this reason we need to change this API function to this one:
```suggestion
  ExportData exportData(1: RunFilter runFilter)
```
For this I created the new packages: `6.37.0-dev5`.

csordasmarton commented on web/server/codechecker_server/api/report_server.py:
```diff
@@ -3279,3 +3279,37 @@ def getAnalysisStatistics(self, run_id, run_history_id):
+
+            # Logic for getting comments
+            comment_data_list = defaultdict(list)
+            comment_query = session.query(Comment, Report.bug_id) \
+                .filter(Report.run_id == run_id) \
+                .outerjoin(Report, Report.bug_id == Comment.bug_hash) \
+                .order_by(Comment.created_at.desc())
```
Handle the use case when no run filters are given.

csordasmarton commented on web/server/codechecker_server/api/report_server.py:
```diff
@@ -3279,3 +3279,37 @@ def getAnalysisStatistics(self, run_id, run_history_id):
+                comment_data_list[report_id].append(comment_data)
+
+            # Logic for getting review status
+            review_data_list = defaultdict(ReviewData)
```
No need to use defaultdict here. We can use a simple dict.

csordasmarton commented on web/server/codechecker_server/api/report_server.py:
```diff
@@ -3279,3 +3279,37 @@ def getAnalysisStatistics(self, run_id, run_history_id):
+                .order_by(ReviewStatus.date.desc())
+
+            for data, report_id in review_query:
+                report_id = report_id
```
```suggestion
```

csordasmarton commented on web/server/codechecker_server/api/report_server.py:
```diff
@@ -3279,3 +3279,37 @@ def getAnalysisStatistics(self, run_id, run_history_id):
+
+            # Logic for getting review status
+            review_data_list = defaultdict(ReviewData)
+            review_query = session.query(ReviewStatus, Report.bug_id) \
+                .filter(Report.run_id == run_id) \
+                .outerjoin(Report, Report.bug_id == ReviewStatus.bug_hash) \
+                .order_by(ReviewStatus.date.desc())
```
Handle the use case when no run filter is given.

csordasmarton reviewed (changes requested):
Sorry for the late reply but I was on holiday. Thank you very much for this patch, I reviewed it and I also had some comments.

I didn't find the IMPORT functionality in this patch and I also saw that you created a separate branch for it (cmd_importer). It would be better to add this functionality in this branch because we already created API functions for it, and if we merge this one the user will not be able to import data just to export it. Also it will be easier to test the whole functionality whether I can export and import data from the server. 
So if it's okay for you could you please cherry pick your commits from your import branch to this branch and I will review those changes to.

Thank you very much for your hard work :blush:

jay24rajput commented:
@csordasmarton Thanks you for your feedback. I have created a separate branch called `cmd_importer` because I did not want to mess the commits on this pull request. Once the functionality is fully available I will push it to this branch maybe. 

Regarding the https://github.com/Ericsson/codechecker/pull/3116#discussion_r553184474 if we export all the reports:
1. The size of json would become really big
2. There would be problems while importing the data since we are using the json file name as the default name of the run, if not explicitly provided by the user. The importer functionality build so far assumes and revolves around this. 

I think it would be good to prompt the user to provide the name of the and import/export reports for a single run

csordasmarton commented:
The review statuses and comments are connected to report hashes (not for report ids) and these are user specified and not every report has comments. So this way I don't think it would be so huge. Also it is possible that I want to see all the comments for a product or to save all the source code comments as a backup or to save data for multiple runs. So I think with small changes we can do this. Maybe for your use case it is enough to export the data for a single run, but we should think to other users and use cases as well.

By default the export command should print the data to the standard output and we can redirect it to a file anytime (and the file name can be a run name or whatever you want). When we importing data to the server the run name is not important, we will import every entry from the exported file.

jay24rajput commented:
@csordasmarton That makes sense. I will update the code as per your suggestions, I will also modify the importer functionality

jay24rajput commented on web/client/codechecker_client/cmd/cmd.py:
```diff
@@ -1430,5 +1445,18 @@ def add_arguments_to_parser(parser):
+    export = subcommands.add_parser(
+        'export',
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+        description="Export the comments and review status from "
+                    "codechecker server into a json format",
```
Doing this fails pycodestyle test cases

csordasmarton replied:
Yes, because you have to break the lines correctly. I didn't break it in this suggestion.

jay24rajput commented:
@csordasmarton I have implemented the changes suggested by you for the exporter and now we can have the exporter export the data from multiple and all( in case of no run is provided).

I have also implemented the functionality for the importer command

csordasmarton commented on web/server/codechecker_server/api/report_server.py:
```diff
@@ -3279,3 +3279,124 @@ def getAnalysisStatistics(self, run_id, run_history_id):
+        with DBSession(self.__Session) as session:
+
+            # Logic for importing comments
+            imported_data = {}
+            imported_comments = []
+
+            for bug_hash, comments in exportData.comments.items():
+                for comment in comments:
+                    imported_data[comment.id] = [bug_hash, comment]
+                    imported_comments.append(comment)
+
+            comment_bug_ids = list(exportData.comments.keys())
+            current_query = session.query(Comment)\
+                .filter(Report.bug_id.in_(comment_bug_ids)) \
+                .outerjoin(Report, Report.bug_id == Comment.bug_hash) \
+                .order_by(Comment.created_at.desc())
+
+            db_comments = {}
+            for comment in current_query:
+                db_comments[comment.id] = comment
+
+            for imported_comment in imported_comments:
+                # Check if the new comment exists in the database
+                if imported_comment.id in db_comments.keys():
+
+                    db_comment = db_comments.get(imported_comment.id)
+                    message = get_comment_msg(db_comment)
+
+                    # Check if the data is same for both
+                    if imported_comment.message != message:
+                        db_comment.message = imported_comment.message.encode(
+                            'utf-8')
+                        session.add(db_comment)
+
+                # If the comment does not exist add it
+                else:
+                    data = imported_data.get(imported_comment.id)
+                    new_comment = self.__add_comment(data[0], data[1].message)
+                    session.add(new_comment)
```
Importing comments to the database is not so easy.
For example we can't use the comment id. Lets suppose that in the 2 users are using different servers. Both user insert a new comment on their own server with different messages. Both comment will have ID-1. If the first user exports the data from the server and send it to the second user who tries to import it, he would like to see both comments. So we can't say if the comment ID's are equal we will just update the comment.
To see if a comment is in the database we can check if the date, kind and messages fields are equals. If not, we will insert a new comment.

What do you think?

```suggestion
                        comment_bug_ids = list(exportData.comments.keys())
            comment_query = session.query(Comment) \
                .filter(Comment.bug_hash.in_(comment_bug_ids)) \
                .order_by(Comment.created_at.desc())

            comments_in_db = defaultdict(list)
            for comment in comment_query:
                comments_in_db[comment.bug_hash].append(comment)

            for bug_hash, comments in exportData.comments.items():
                db_comments = comments_in_db[bug_hash]
                for comment in comments:
                    date = datetime.strptime(comment.createdAt,
                                             '%Y-%m-%d %H:%M:%S.%f')

                    message = comment.message.encode('utf-8') \
                        if comment.message else b''

                    # See if the comment is already in the database.
                    if any(c.created_at == date and \
                           c.kind == comment.kind and \
                           c.message == message for c in db_comments):
                        continue

                    c = Comment(bug_hash, comment.author, message,
                                comment.kind, date)
                    session.add(c)
```

jay24rajput replied:
Yes, this is certainly a better approach over just comparing `id` and `message`. I had followed the suggestion given by you in this comment thread: https://github.com/Ericsson/codechecker/issues/3076#issuecomment-747300763

csordasmarton commented on web/server/codechecker_server/api/report_server.py:
```diff
@@ -3279,3 +3279,124 @@ def getAnalysisStatistics(self, run_id, run_history_id):
+
+            # Logic for importing review status
+            review_bug_ids = list(exportData.reviewData.keys())
+            review_query = session.query(Report.bug_id, ReviewStatus,
+                                         Report.id)\
+                .filter(Report.bug_id.in_(review_bug_ids)) \
+                .outerjoin(Report, Report.bug_id == ReviewStatus.bug_hash) \
+                .order_by(ReviewStatus.date.desc())
+
+            review_data = {}
+            for bug_hash, review, report_id in review_query:
+                review_data[bug_hash] = [review, report_id]
+
+            # print(review_data)
+
+            for bug_hash, imported_review in exportData.reviewData.items():
+
+                # If the review does not exist in the database
+                if bug_hash not in review_data:
+                    updated_status = imported_review.status
+                    updated_message = imported_review.comment
+
+                    report_id = session.query(Report.id)\
+                        .filter(Report.bug_id == bug_hash)\
+                        .all()
+
+                    report_id = report_id[0][0]
+
+                    res = self._setReviewStatus(
+                        report_id, updated_status, updated_message, session)
+
+                # Check if the date of the report is same
+                elif (str(review_data[bug_hash][0].date)
+                      != imported_review.date):
+                    updated_status = imported_review.status
+                    updated_message = imported_review.comment
+                    report_id = review_data[bug_hash][1]
+
+                    res = self._setReviewStatus(
+                        report_id, updated_status, updated_message, session)
```
Here we get the report id because the `_setReviewStatus` requires it. First I recommend you to rebase your branch because there was an update in this function. Next we should change the parameter from report id to report hash: `def _setReviewStatus(self, session, report_hash, status, message, date=None):`. And there are already two places where we call this API function. The first place we have the `bug_id` what we can pass to this function so there is no need to get the report from the database. The other place we need to get the report from the database by report id and then pass the bug hash of this report to the set review status function:
https://github.com/Ericsson/codechecker/blob/b4f569dfa7b989424baf7876a96696cff38a7a55/web/server/codechecker_server/api/report_server.py#L1644

This way we can simplify this code on the following way:
```suggestion
                        review_bug_ids = list(exportData.reviewData.keys())
            review_query = session.query(ReviewStatus) \
                .filter(ReviewStatus.bug_hash.in_(review_bug_ids)) \
                .order_by(ReviewStatus.date.desc())

            db_review_data = {}
            for review_status in review_query:
                db_review_data[review_status.bug_hash] = review_status

            for bug_hash, imported_review in exportData.reviewData.items():
                db_status = db_review_data.get(bug_hash)

                # The status is up-to-date.
                if db_status and str(db_status.date) == imported_review.date:
                    continue

                date = datetime.strptime(imported_review.date,
                                         '%Y-%m-%d %H:%M:%S.%f')

                self._setReviewStatus(session,
                                      bug_hash,
                                      imported_review.status,
                                      imported_review.comment,
                                      date)
```

csordasmarton commented on web/tests/functional/export/test_export.py:
```diff
@@ -0,0 +1,116 @@
+        logging.debug('Bug review status changed successfully')
+
+        # Export the comments and reviews
+        exported = self._cc_client.exportData(run_filter)
```
Here you test only the export feature. Can we test the import feature too somehow?

For example at the end of this test case we remove some comments and then we can import it again.

jay24rajput replied:
Hmm. I think this can be done, but then would you suggest changing the same of the root directory of this folder. Cause as of now, it is just named as `importer`.

Another approach would be making a separate folder for the importer and doing all the steps done for the importer and actually importing actual json file, but IMO that will be much harder. 

I think we would go with the first approach itselt, what do you think @csordasmarton ??

csordasmarton replied:
For now I think it will be good to do the following:
- Rename the root folder to `export_import`.
- Rename the test file to `test_export_import.py`.
- Rename the test case from `test_export` to `test_export_import`.
- What you did in this test case is good. The next thing we should do in this test case is to export the data to a json file similar to this: https://github.com/Ericsson/codechecker/blob/dd194907919f020b2b825bf1f66f2b544e855db4/web/tests/functional/diff_local/test_diff_local.py#L74-L80. So you have to call the CodeChecker cmd export command and print the output to a temporary json file. For example you can create a temp file like this: https://github.com/Ericsson/codechecker/blob/dd194907919f020b2b825bf1f66f2b544e855db4/web/tests/functional/cmdline/test_cmdline.py#L268
- Next remove the first comment by using the `removeComment` API function.
- Update the message of the second comment by using the `updateComment` API function.
- Call the CodeChecker cmd import function with the temporary json file created above.
- Check whether the data is imported to the server. You can use the `exportData` method similar what you did before and see if it contains results from the import set or you can use the `getComments` API function and check the results.

jay24rajput replied:
@csordasmarton writing it to a json file would require the formatting for again converting the output from json into exportData format(https://github.com/jay24rajput/codechecker/blob/e47a86d23cb9ca2454cfd9c561c5a543b07330ff/web/client/codechecker_client/cmd_line_client.py#L1681) whereas here we already have an exportData object exported(https://github.com/jay24rajput/codechecker/blob/e47a86d23cb9ca2454cfd9c561c5a543b07330ff/web/tests/functional/export/test_export.py#L105) from which it is easy to delete/modify comments and review status and again create a new exportData object directly. I think for the testing purpose writing it to a json format and again importing it would be kind of redundant since we are testing the `importdata` method which only takes an exportData object as an input

csordasmarton replied:
Yes, you are right. Here we have the data but I recommend to use the commands (CodeChecker cmd export/import) instead of calling API functions. The test will more robust, because on the client side commands we do a couple of extra things. And most of the user will use these commands instead of the API functions. And with this we would test whether the conversion to json works properly, it is a valid json file, we can import this json file to the server etc.

jay24rajput replied:
@csordasmarton In the following code if executed for the export_import, it shows error in the subprocess.check command.
https://github.com/jay24rajput/codechecker/blob/cmd_importer/web/tests/functional/export_import/test_export_import.py

I am attaching the stack trace of the same:
```
ERROR: Test exporting and importing feature
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jay/codechecker/web/tests/functional/export_import/test_export_import.py", line 122, in test_export
    out_json = subprocess.check_output(new_export_command, encoding="utf-8", errors="ignore")
  File "/usr/lib/python3.8/subprocess.py", line 411, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.8/subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/home/jay/codechecker/build/CodeChecker/bin/CodeChecker', 'cmd', 'export', '-n', 'test_files_5f78c3e2b3c34bfcbcade4c2eb1af76e']' returned non-zero exit status 1.
-------------------- >> begin captured stdout << ---------------------
Running TestExport tests in /home/jay/codechecker/build/workspace/export_import-3_0qq4zf
Could not load session for session getter because 'localhost:45537'
Setup viewer client: http://localhost:45537/export_import/v6.37/CodeCheckerService
['/home/jay/codechecker/build/CodeChecker/bin/CodeChecker', 'cmd', 'export', '-n', 'test_files_5f78c3e2b3c34bfcbcade4c2eb1af76e']
```

May you know why is error is occurring? Cause the syntax for running the command is:
```sh
CodeChecker cmd export -n <run_name>
```
May you know why this error? If we comment out the lines suggested by you the test runs successfully

csordasmarton replied:
Sorry, my bad. I gave you a wrong example. In our case you have to specify the `--url` option to this command too:
https://github.com/Ericsson/codechecker/blob/9d23b7ba6d4c149a6fabf21579f08fc60ef4df1d/web/tests/functional/cmdline/test_cmdline.py#L251

In the `setUp` function you can use this:
https://github.com/Ericsson/codechecker/blob/9d23b7ba6d4c149a6fabf21579f08fc60ef4df1d/web/tests/functional/cmdline/test_cmdline.py#L50-L52

jay24rajput replied:
@csordasmarton Yes that seems to work! Thanks a lot! For the import command I am planning to do something like:
```python
 with tempfile.NamedTemporaryFile() as component_f:
            component_f.write(out_json.encode('utf-8'))
           # Will delete and update the comments here in codechecker server

            new_import_command = [self._codechecker_cmd, 'cmd', 'import',
                                  '-i', component_f.name,
                                  '--url', str(self.server_url)]
            print(new_import_command)
            output = subprocess.check_output(
                new_export_command, encoding="utf-8", errors="ignore")
```

Does this method seem right to you? Or you have any other method of doing the same?

csordasmarton replied:
It looks good except that in the last line you are calling the export command (`new_export_command`) and not the import (`new_import_command`) :blush:

jay24rajput replied:
Sadly even this is returning the following error:
```
ERROR: Test exporting and importing feature
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jay/codechecker/web/tests/functional/export_import/test_export_import.py", line 145, in test_export
    output = subprocess.check_output(
  File "/usr/lib/python3.8/subprocess.py", line 411, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.8/subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/home/jay/codechecker/build/CodeChecker/bin/CodeChecker', 'cmd', 'import', '-i', '/tmp/tmpa5nvjjzi', '--url', 'localhost:43645/export_import']' returned non-zero exit status 
```

csordasmarton replied:
I think the problem will be that you have to `flush` the output after calling the `write` method:
```py
component_f.write(out_json.encode('utf-8'))
component_f.flush()
component_f.seek(0)
````

Also you can catch these error with a simple try/except blocks:
```py
try:
    output = subprocess.check_output(
        new_import_command, encoding="utf-8", errors="ignore")
except subprocess.CalledProcessError as e:
    print(e.output)
```

Also if you have such problem, please don't forget to update your branch because it will be easier for me to test it.

jay24rajput commented on web/client/codechecker_client/cmd/cmd.py:
```diff
@@ -1430,5 +1458,28 @@ def add_arguments_to_parser(parser):
+    __add_common_arguments(export, output_formats=['json'])
+
+    importer = subcommands.add_parser(
+        'import',
```
Yes, Let me issue a separate commit only for updating the documentation of these commands

jay24rajput commented on web/server/codechecker_server/api/report_server.py:
```diff
@@ -3279,3 +3279,124 @@ def getAnalysisStatistics(self, run_id, run_history_id):
+                .order_by(Comment.created_at.desc())
+
+            for data, report_id in comment_query:
+                comment_data = comment_data_db_to_api(data)
```
Yes, I agree. Although thought it would be better for the user to understand the comments in the outputted json file if we format it. But yes, that would require extra processing on both importer and exporter side

jay24rajput commented:
@csordasmarton @bulwahn I hope I have addressed all the changes and suggestions given by you. I have also added the docs as instructed by you Kindly have a look 😄

csordasmarton commented on docs/web/user_guide.md:
```diff
@@ -37,6 +37,8 @@ Table of Contents
         * [`products` (Manage product configuration of a server)](#cmd-product)
         * [`login` (Authenticate to the server)](#cmd-login)
+        * [`export` (Export comments and review statuses from CodeChecker)](#cmd-export)
+        * [`import` (Import comments and review statuses into CodeChecker)`](#cmd-import)
```
```suggestion
        * [`import` (Import comments and review statuses into CodeChecker)](#cmd-import)
```

csordasmarton commented on docs/web/user_guide.md:
````diff
@@ -1579,6 +1583,49 @@ not found, the user will be asked, in the command-line, to provide credentials.
+### Export comments and review statuses (`export`) <a name="cmd-export"></a>
+
+```sh
+usage: CodeChecker cmd export [-h] [-n RUN_NAME [RUN_NAME ...]] [--url PRODUCT_URL] [--verbose {info,debug_analyzer,debug}]
+
+Export data (comments, review statuses) from a running CodeChecker server into a json format
+
+optional arguments:
+  -h, --help            show this help message and exit
+  -n RUN_NAME [RUN_NAME ...], --name RUN_NAME [RUN_NAME ...]
+                        Name of the analysis run.
+
+common arguments:
+  --url PRODUCT_URL     The URL of the product which will be accessed by the client, in the format of '[http[s]://]host:port/Endpoint'. (default:
+                        localhost:8001/Default)
+  --verbose {info,debug_analyzer,debug}
+                        Set verbosity level.
````
We do not need to list `common arguments` here. We don't list it for the other commands too.

Also please don't remove line breaks. Just run this command `CodeChecker cmd export -h` and copy-paste the output to here without the common arguments section :blush: 

```suggestion
usage: CodeChecker cmd export [-h] [-n RUN_NAME [RUN_NAME ...]]
                              [--url PRODUCT_URL]
                              [--verbose {info,debug_analyzer,debug}]

Export data (comments, review statuses) from a running CodeChecker server into
a json format

optional arguments:
  -h, --help            show this help message and exit
  -n RUN_NAME [RUN_NAME ...], --name RUN_NAME [RUN_NAME ...]
                        Name of the analysis run.
```

csordasmarton commented on docs/web/user_guide.md:
````diff
@@ -1579,6 +1583,49 @@ not found, the user will be asked, in the command-line, to provide credentials.
+
+### Import comments and review statuses into Codechecker (`import`) <a name="cmd-import"></a>
+```sh
+usage: CodeChecker cmd import [-h] -i JSON_FILE [--url PRODUCT_URL] [--verbose {info,debug_analyzer,debug}]
+
+Import the results into CodeChecker server
+
+optional arguments:
+  -h, --help            show this help message and exit
+  -i JSON_FILE, --import JSON_FILE
+                        Import findings from the json file into the database.
+
+common arguments:
+  --url PRODUCT_URL     The URL of the product which will be accessed by the client, in the format of '[http[s]://]host:port/Endpoint'. (default:
+                        localhost:8001/Default)
+  --verbose {info,debug_analyzer,debug}
+                        Set verbosity level.
````
Same as my above comment:
```suggestion
usage: CodeChecker cmd import [-h] -i JSON_FILE [--url PRODUCT_URL]
                              [--verbose {info,debug_analyzer,debug}]

Import the results into CodeChecker server

optional arguments:
  -h, --help            show this help message and exit
  -i JSON_FILE, --import JSON_FILE
                        Import findings from the json file into the database.
```

csordasmarton commented on web/client/codechecker_client/cmd/cmd.py:
```diff
@@ -1430,5 +1458,28 @@ def add_arguments_to_parser(parser):
+    export = subcommands.add_parser(
+        'export',
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+        description="Export data (comments, review statuses) from a running "
+        "CodeChecker server into a json format",
```
Do not we need to indent this text here?
```suggestion
        description="Export data (comments, review statuses) from a running "
                    "CodeChecker server into a json format",
```

csordasmarton commented on web/client/codechecker_client/cmd/cmd.py:
```diff
@@ -1430,5 +1458,28 @@ def add_arguments_to_parser(parser):
+        'import',
+        formatter_class=arg.RawDescriptionDefaultHelpFormatter,
+        description="Import the results into CodeChecker server",
+        help="Import the analysis from a json file exported by the "
+        "'CodeChecker cmd export' command into CodeChecker"
```
Same as above:
```suggestion
        help="Import the analysis from a json file exported by the "
             "'CodeChecker cmd export' command into CodeChecker"
```

csordasmarton commented on web/client/codechecker_client/cmd_line_client.py:
```diff
@@ -1643,3 +1643,63 @@ def handle_list_run_histories(args):
+                                   reviewData=review_data_list)
+
+    status = client.importData(exportData)
+    if status:
+        LOG.info("Import successful!")
```
We need to handle the use case when the import is failed:
```suggestion
    if not status:
        LOG.error("Failed to import data!")
        sys.exit(1)

    LOG.info("Import successful!")
    
```

csordasmarton commented on web/server/codechecker_server/api/report_server.py:
```diff
@@ -1549,14 +1549,16 @@ def getReportDetails(self, reportId):
         needed because during storage a specific session object has to be used.
         """
-        report = session.query(Report).get(report_id)
+        report = session.query(Report).filter(Report.bug_id == report_hash)\
+            .all()[0]
```
Review statuses are connected to report hashes. Previously we fetched the report from the database because the report id was given to this function and we needed the report hash for this report. Now we have the report hash in the parameter so there is no need to fetch the report here.
You can remove this line the if condition and the else branch and indent the code like this in the if branch:
```suggestion
        review_status = session.query(ReviewStatus).get(report_hash)
        ...
```

csordasmarton commented on web/server/codechecker_server/api/report_server.py:
```diff
@@ -1641,7 +1643,9 @@ def changeReviewStatus(self, report_id, status, message):
 
         with DBSession(self.__Session) as session:
-            res = self._setReviewStatus(session, report_id, status, message)
+            report = session.query(Report).get(report_id)
+            res = self._setReviewStatus(
+                session, report.bug_id, status, message)
```
We need to handle the use case when no report is found in the database:
```suggestion
            report = session.query(Report).get(report_id)
            if report:
                res = self._setReviewStatus(
                    session, report.bug_id, status, message)
            else:
                raise codechecker_api_shared.ttypes.RequestFailed(
                    codechecker_api_shared.ttypes.ErrorCode.DATABASE,
                    "No report found in the database.")
```

csordasmarton reviewed (changes requested):
I have some other tiny comments. Once these are fixed I will test it again and I think we can merge it if everything works properly :blush:

jay24rajput commented:
@csordasmarton I have made changes as per your suggestions, kindly have a look

csordasmarton reviewed (approved):
LGTM! Thank you very much for this contribution :blush:
